infra/relayer/src/purse.js

```diff
@@ -0,0 +1,682 @@
+    return await this.web3.eth.accounts.signTransaction(
+      tx,
+      bufferToHex(privKey)
+    )
+  }
+
+  /**
+   * Send a transaction from an available sender account.
+   *
+   * NOTE: This does not wait for the transaction to be mined!
+   *
+   * @param tx {object} - The transaction object, sans `from` and `nonce`
+   * @param onReceipt {function} - An optional callback to call when a receipt is found
+   * @returns {string} The transaction hash of the sent transaction
+   */
+  async sendTx(tx, onReceipt) {
+    const address = await this.getAvailableAccount()
+
+    // Set the from and nonce for the account
+    tx = {
+      ...tx,
+      from: address,
+      nonce: await this.txCount(address)
+    }
+
+    if (!tx.gasPrice) {
+      const gasPrice = stringToBN(await this.web3.eth.getGasPrice())
+      if (gasPrice.gt(MAX_GAS_PRICE)) {
+        // TODO: best way to handle this?
+        throw new Error('Current gas prices are too high!')
+      }
+      tx.gasPrice = gasPrice
+    }
+
+    logger.debug('Sending tx: ', tx)
+
+    const signed = await this.signTx(address, tx)
+    const rawTx = signed.rawTransaction
+    // Not in this version of web3.js 1.0...
+    // const txHash = signed.transactionHash
+    const txHash = this.web3.utils.sha3(rawTx)
+
+    // If there's an onReceipt cb, store it for later
+    if (onReceipt) {
+      if (typeof onReceipt !== 'function') {
+        throw new Error('onReceipt is not a function')
+      }
+      this.receiptCallbacks[txHash] = onReceipt
+    }
+
+    // In case it needs to be rebroadcast
+    this.pendingTransactions[txHash] = rawTx
+    await this.incrementTxCount(address)
+
+    // blast it
+    await sendRawTransaction(this.web3, rawTx)
+
+    logger.info(`Sent ${txHash}`)
+
+    return txHash
+  }
+
+  /**
+   * Get the current transaction count for an account (includes known pending)
+   * @param address {string} - The account address to increment
+   */
+  async txCount(address) {
+    let txCount = this.accounts[address] ? this.accounts[address].txCount : 0
+
+    // If we have one in memory, go with that
+    if (txCount > 0) return txCount
+
+    // Check redis, if available
+    if (txCount === 0 && this.rclient && this.rclient.connected) {
+      const countFromRedis = await this.rclient.getAsync(
+        `${REDIS_TX_COUNT_PREFIX}${address}`
+      )
+      // null defense
+      if (countFromRedis) {
+        txCount = countFromRedis
+      }
+    } else {
+      logger.warn('Redis unavailable')
+    }
+
+    /**
+     * Finally, check with the JSON-RPC provider.  Chance of some pendings screwing this up if we
+     * get this far.  I'm at least sure that eth_pendingTransactions, is completely  not supported
+     * by Alchemy.  Doubtful from Infura as well.  And it's a poor method anyway with no arguments.
+     *
+     * EDGE CASE WARNING
+     * -----------------
+     *
+     * The only time this should be an issue is if the service is magically restarted, Redis is
```

Comment:
Yea was going to ask what happens if the service is restarted, for example when we do a new deployment and there are still transactions pending

Reply:
To be clear, the edge case I mention here isn't triggered just by a service restart.  A regular old service restart shouldn't have any negative effect.